hutch_python/ipython_log.py

```diff
@@ -22,47 +43,157 @@ class IPythonLogger:
 
     Parameters
     ----------
-    ipython: ``ipython`` ``Shell``
+    ipython : ``IPython.terminal.interactiveshell.TerminalInteractiveShell``
         The active ``ipython`` ``Shell``, perhaps the one returned by
         ``IPython.get_ipython()``.
+
+    Attributes
+    ----------
+    ipython_in : list of str
+        The IPython user input list.
+
+    prev_err_value : Exception or None
+        The last exception value that was logged.  Used for exception
+        deduplication.
+
+    line_in_progress : bool
+        True if a line is currently in the process of being evaluated.
     """
     def __init__(self, ipython):
-        self.prev_err = None
-        self.In = ipython.user_ns['In']
-        self.Out = ipython.user_ns['Out']
+        self.ipython = ipython
+        self.ipython_in = ipython.user_ns["In"]
+        self.prev_err_value = None
+        self.line_in_progress = False
+
+        logging.addLevelName('INPUT', INPUT_LEVEL)
+
+        ipython.events.register('pre_run_cell', self.log_user_input)
+        ipython.events.register('post_run_cell', self.log_output)
+
+        self._orig_sys_excepthook = sys.excepthook
+        sys.excepthook = self._sys_exception_hook
```

Comment:
sys.excepthook seems to do nothing if running in IPython

Reply:
That reflects what I saw, too. I'm not sure how you're supposed to catch background thread exceptions pre-3.8, though, if `excepthook` doesn't do anything. A bit confusing...

Reply:
From some reading around stackoverflow, there is some alternate handling you can do for basic excepthook things: https://stackoverflow.com/a/28758396

But this is ugly and I'm not sure this even helps is a way at all to catch background thread exceptions before py3.8

Good thing we're at py38